### PR TITLE
Library refactor

### PR DESCRIPTION
**Pull request by SeanCampbell**
A bit of refactoring so that reading the config file is done outside of main and global variables are separated into their own files.

### CODE REVIEW

File: Makefile.am [hunk 1]
```diff
@@ -199,6 +199,8 @@ collectd_SOURCES = \
 	src/daemon/configfile.h \
 	src/daemon/filter_chain.c \
 	src/daemon/filter_chain.h \
+        src/daemon/globals.c \
```
Comment (octo) [inline Makefile.am]: Nit: please fix whitespace. Looks like a tabs vs. spaces issue.

File: src/daemon/collectd.c [hunk 2]
```diff
@@ -454,21 +444,17 @@ static int notify_systemd(void) {
...
-  int test_config = 0;
-  int test_readall = 0;
-  const char *basedir;
-#if COLLECT_DAEMON
-  pid_t pid;
-  int daemonize = 1;
-#endif
-  int exit_status = 0;
+struct cmdline_config {
+  int test_config;
```
Comment (octo) [inline src/daemon/collectd.c]: Boolean fields should be of type `_Bool`.

**Comment by SeanCampbell** [comment]
Hi @octo,

Thanks for your patience. I tidied up the PR so it's just focused on the refactor.

Let me know if there's any other changes I should make :)

Thanks,
-Sean

**Comment by octo** [comment]
Hi @SeanCampbell,

thanks for the update! Unfortunately you merged *master* into your PR, which makes code reviews confusing. I pushed a cleaned-up branch of your changes to [octo/collectd pr/2467](https://github.com/collectd/collectd/compare/master...octo:pr/2467), can we use that to discuss this change? I'm happy to push it into the PR branch, or you could do something along these lines:

```sh
# Make sure you're on the right branch
git checkout library-refactor
# Download my commits
git fetch git@github.com:octo/collectd.git pr/2467
# Reset the "library-refactor" branch to point to the latest commit
git reset --hard FETCH_HEAD
# Force pushing of the new head to the tracking remote
git push -f
```

Best regards,
—octo

**Comment by SeanCampbell** [comment]
Ah, it said there were conflicts that prevented it from being automatically merged, so I thought that would make it easier. I ran the commands you said to pull in the updates from your branch, and happy to make any changes :)

Thank you,
Sean

File: src/daemon/globals.c [hunk 3]
```diff
@@ -0,0 +1,13 @@
+#include "globals.h"
+#include "plugin.h"
+/*
+ * Global variables
+ */
+char hostname_g[DATA_MAX_NAME_LEN];
+const int hostname_g_size = sizeof (hostname_g);
+cdtime_t interval_g;
+int  pidfile_from_cli = 0;
```
Comment (octo) [inline src/daemon/globals.c]: `pidfile_from_cli` is unused, likely a remnant of an unrelated change. Please remove.

File: src/daemon/globals.c [hunk 4]
```diff
@@ -0,0 +1,13 @@
+#include "globals.h"
```
Comment (octo) [inline src/daemon/globals.c]: This file needs a license header. In case of the core daemon, we insist on the [MIT license](https://opensource.org/licenses/MIT).

Dito for `src/daemon/globals.h`.

File: src/daemon/globals.c [hunk 5]
```diff
@@ -0,0 +1,13 @@
...
+#include "plugin.h"
+/*
+ * Global variables
+ */
+char hostname_g[DATA_MAX_NAME_LEN];
+const int hostname_g_size = sizeof (hostname_g);
```
Comment (octo) [inline src/daemon/globals.c]: I'm not quite happy about the introduction of this variable. How about we change `hostname_g` to be a `char *` instead? I think it would also be a good idea to add a simple setter function:

```c
/* hostname_set updates hostname_g */
void hostname_set(char const *hostname);
```

File: src/daemon/globals.c [hunk 6]
```diff
@@ -0,0 +1,13 @@
+#include "globals.h"
+#include "plugin.h"
```
Comment (octo) [inline src/daemon/globals.c]: I think you need `"plugin.h"` for `DATA_MAX_NAME_LEN`, which is one of those defines that might actually make sense to move to here (and out of `"plugin.h"`).

File: src/daemon/globals.c [hunk 7]
```diff
@@ -0,0 +1,13 @@
...
+/*
+ * Global variables
+ */
+char hostname_g[DATA_MAX_NAME_LEN];
+const int hostname_g_size = sizeof (hostname_g);
+cdtime_t interval_g;
+int  pidfile_from_cli = 0;
+int  timeout_g;
+#if HAVE_LIBKSTAT
+kstat_ctl_t *kc;
```
Comment (octo) [inline src/daemon/globals.c]: This type is not defined because the header is missing. I recommend to move

```c
#if HAVE_KSTAT_H
#include <kstat.h>
#endif
```

out of `src/daemon/collectd.h` and into here. Longer term, I'd like to get rid of the `kc` variable, but that's out of scope for here.

File: src/daemon/globals.c [hunk 8]
```diff
@@ -1,13 +1,38 @@
...
+ * FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER
+ * DEALINGS IN THE SOFTWARE.
+ **/
+
+#include "common.h"
 #include "globals.h"
-#include "plugin.h"
+
+void hostname_set(char const *hostname) {
+  sstrncpy(hostname_g, hostname, sizeof(hostname_g));
```
Comment (octo) [inline src/daemon/globals.c]: `sizeof()` only works with `char[]` (array) but not with `char*` (pointer).

I'd recommend doing something like this:

```c
void hostname_set(char const *hostname) {
  char *h = strdup(hostname);
  if (h == NULL)
    return;

  free(hostname_g);
  hostname_g = h;
}

File: src/daemon/collectd.c [hunk 9]
```diff
@@ -83,11 +83,11 @@ static int init_hostname(void) {
 
   str = global_option_get("Hostname");
   if ((str != NULL) && (str[0] != 0)) {
-    sstrncpy(hostname_g, str, hostname_g_size);
+    hostname_set(str);
     return 0;
   }
 
-  if (gethostname(hostname_g, hostname_g_size) != 0) {
+  if (gethostname(hostname_g, sizeof(hostname_g)) != 0) {
```
Comment (octo) [inline src/daemon/collectd.c]: With `hostname_g` being a `char*`, you need to use a temporary buffer here. Sizing is a bit tricky, but the correct way would be:

```c
long hostname_len = sysconf(_SC_HOST_NAME_MAX);
if (hostname_len == -1) {
  hostname_len = NI_MAXHOST;
}
char hostname[hostname_len];
```

Then use `hostname` instead of `hostname_g` for the remainder of the function.

P.S.: If the below `getaddrinfo()` call fails, you'll have to add a `hostname_set(hostname);` to keep the existing behavior of "configured-but-unresolvable is better than no hostname".

File: src/daemon/collectd.h [hunk 10]
```diff
@@ -186,10 +186,6 @@
 #include <sys/param.h>
 #endif
 
-#if HAVE_KSTAT_H
```
Comment (octo) [inline src/daemon/collectd.h]: `"collectd.h"` used to be the header defining the global variables, so it should include `"globals.h"` now so that plugins continue working / compiling.

File: src/daemon/globals.c [hunk 11]
```diff
@@ -1,13 +1,38 @@
+/**
+ * collectd - src/globals.c
+ * Copyright (C) 2017  Florian octo Forster
```
Comment (octo) [inline src/daemon/globals.c]: I think the OSPO would want you to put "Google, Inc." here. Dito for `globals.h`.

Comment (SeanCampbell) [reply src/daemon/globals.c]: Gotcha. I put "Google LLC" because of the recent name change. Is that right?

**Comment by SeanCampbell** [comment]
Fixed the issue with hostname_g[] (for some reason, it was compiling fine on my machine even though it broke the tests). I'm happy to take any other comments or suggestions.

File: src/daemon/collectd.h (hunk 10 above)
Comment (SeanCampbell) [reply src/daemon/collectd.h]: Yes, it's included at the bottom :)

**Comment by SeanCampbell** [comment]
Friendly ping :)

File: src/daemon/collectd.c [hunk 12]
```diff
@@ -91,13 +81,20 @@ static int init_hostname(void) {
   struct addrinfo *ai_list;
   int status;
 
+  long hostname_len = sysconf(_SC_HOST_NAME_MAX);
+  if (hostname_len == -1) {
+    hostname_len = NI_MAXHOST;
+  }
+  char hostname[hostname_len];
+  hostname_set(hostname);
```
Comment (octo) [inline src/daemon/collectd.c]: You're calling `hostname_set()` with an uninitialized buffer here. I think this needs to be removed – the code below is taking care of calling `hostname_set()` once the buffer is initialized.

File: src/daemon/globals.c [hunk 13]
```diff
@@ -0,0 +1,49 @@
...
+char *hostname_g;
+cdtime_t interval_g;
+int  timeout_g;
+#if HAVE_KSTAT_H
+kstat_ctl_t *kc;
+#endif
+
+void hostname_set(char const *hostname) {
+    char *h = strdup(hostname);
+      if (h == NULL)
```
Comment (octo) [inline src/daemon/globals.c]: The indentation here is wrong. Please run `clang-format -style=file -i $file` on each `.c` and `.h` file that was changed or added.

File: src/daemon/globals.c [hunk 14]
```diff
@@ -0,0 +1,49 @@
...
+            return;
+
+        free(hostname_g);
```
Comment (octo) [inline src/daemon/globals.c]: Nit: Please use the `sfree()` macro for consistency with other code.

File: src/perl.c [hunk 15]
```diff
@@ -2390,6 +2384,11 @@ static void xs_init(pTHX) {
    * accessing any such variable (this is basically the same as using
    * tie() in Perl) */
   /* global strings */
+  struct {
```
Comment (octo) [inline src/perl.c]: Why did this need to be moved?

Comment (SeanCampbell) [reply src/perl.c]: It was breaking the integration tests with the error message [1]:

src/perl.c:268:1: error: initializer element is not constant
 } g_strings[] = {{"Collectd::hostname_g", hostname_g}, {"", NULL}};
 ^

Since it was a global variable, it could not be initialized with hostname_g as type char*. Changing it to a local variable fixed the issue, and I didn't see anywhere else it was used.

On that note though, I suppose the "g_" prefix would no longer be appropriate. What do you think?

[1] https://ci.collectd.org/job/pull-requests-build-on-jessie-amd64-with-default-toolchain/1766/console

**Comment by SeanCampbell** [comment]
For some reason, after I ran clang-format, it's complaining about the formatting. Any ideas on why that is?

I tried running clang-format on all the files again, just to see if it makes a difference, but it didn't.

**Comment by octo** [comment]
@SeanCampbell Which version of *clang-format* were you using?

**Comment by octo** [comment]
LGTM, thanks @SeanCampbell!

I'd really be interested in which *clang-format* version you were using. The formatting check is brand-new and it's supposed to be helpful, not annoying.

Since we're talking about this: as a contributor, would you prefer if the bot would push the formatted files to the branch instead of reporting a test failure?

Best regards,
—octo

**Comment by SeanCampbell** [comment]
Yeah, I'd like that better, especially since it would eliminate the possibility of differences in clang-format versions causing issues.

Thanks!
-Sean

**Comment by octo** [comment]
FTR, you can use the same *clang-format* the bot is using via cURL:

```
curl --data-binary '@-' https://clang-format.appspot.com <input.c >output.c
```

I'll add a wrapper around this to `contrib/` in a bit.
